data_structures/sparse_table.cpp

```diff
@@ -0,0 +1,141 @@
+#include <array>     /// for std::array
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+/**
+ *
+ * @file
+ * @brief Implementation of Sparse Table
+ *
+ * @author [manncodes](https://github.com/manncodes)
+ *
+ * @details
+ * Sparse Table is a data structure, that allows answering range queries.
+ * It can answer most range queries in O(logn), but its true power is answering
+ * range minimum queries (or equivalent range maximum queries). For those
+ * queries it can compute the answer in O(1) time. The only drawback of this
+ * data structure is, that it can only be used on immutable arrays. This means,
+ * that the array cannot be changed between two queries.
+ *
+ * If any element in the array changes, the complete data structure has to be
+ * recomputed.
+ *
+ * @warning this sparse table is made for min(a1,a2,...an) duplicate invariant
+ * function. This implementation can be changed to other functions like
+ * gcd(),lcm(),max() by changing few lines of code.
+ */
+
+/**
+ * @namespace data_structures
+ * @brief Data Structures algorithms
+ */
+namespace data_structures {
+
+/**
+ * @namespace sparse_table
+ * @brief Implementation of Sparse Table for function : min()
+ *
+ */
+namespace sparse_table {
+
+/**
+ * @brief A struct to represent sparse table
+ *
+ */
+struct Sparse_table {
+    // N : the maximum size of the array.
+    // M : ceil(log2(N)).
+    const static int N = 12345, M = 14;
+
+    // The array to compute its sparse table.
+    int n;
+    std::array<int, N> a;
+
+    //
+    // Sparse table related variables.
+    //
+    // ST[j][i] : the sparse table value (min, max, ...etc) in the interval [i,
+    // i + (2^j) - 1]. LOG[i]   : floor(log2(i)).
+    std::array<std::array<int, N>, M> ST;
+    std::array<int, N> LOG;
```

Comment:
Add a one-line description of these variables (see the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program for more information).